mcs/class/System/System.Net.Sockets/SafeSocketHandle.cs

```diff
@@ -55,20 +56,24 @@ protected override bool ReleaseHandle ()
 						break;
 					}
 
-					/*
-					 * This method can be called by the DangerousRelease inside RegisterForBlockingSyscall
-					 * When this happens blocking_threads contains the current thread.
-					 * We can safely close the socket and throw SocketException in RegisterForBlockingSyscall
-					 * before the blocking system call.
-					 */
 					lock (blocking_threads) {
+						/*
+						* This method can be called by the DangerousRelease inside RegisterForBlockingSyscall
+						* When this happens blocking_threads contains the current thread.
+						* We can safely close the socket and throw SocketException in RegisterForBlockingSyscall
+						* before the blocking system call.
+						*/
 						if (blocking_threads.Count == 1 && blocking_threads[0] == Thread.CurrentThread)
 							break;
-					}
 
-					AbortRegisteredThreads ();
-					// Sleep so other threads can resume
-					Thread.Sleep (1);
+						// abort registered threads
+						foreach (var t in blocking_threads)
+							Socket.cancel_blocking_socket_operation (t);
+
+						// Sleep so other threads can resume
+						in_cleanup = true;
+						Monitor.Wait (blocking_threads, 1000);
```

Comment:
ABORT_RETRIES is now set to 10.
The current code throws an exceptions if fails stop the system call after 10 times, with at least 1ms between them. So we basically have a bit more than 10ms before it fails.
The proposed code may fail when there are more than 10 blocking system calls. This is changing one issue for another.

I think we can fix this by doing something like.
```
var count = blocking_threads.Count;
for(var i < count) {
    if(!Monitor.Wait (blocking_threads, 1000))
         break;
}
```